boefjes/plugins/kat_nmap/normalize.py

```diff
@@ -16,48 +16,54 @@
-
-    ip = (
-        IPAddressV4(network=internet.reference, address=host.address)
-        if host.ipv4
-        else IPAddressV6(network=internet.reference, address=host.address)
-    )
-    yield ip
+def get_ip_ports_and_service(host: NmapHost, network: Network) -> Iterator[OOI]:
+    open_ports = host.get_open_ports()
+    if open_ports:
+        ip = (
+            IPAddressV4(network=network.reference, address=host.address)
+            if host.ipv4
+            else IPAddressV6(network=network.reference, address=host.address)
+        )
+        yield ip
 
-    for port, protocol in host.get_open_ports():
-        service: NmapService = host.get_service(port, protocol)
+        for port, protocol in open_ports:
+            service: NmapService = host.get_service(port, protocol)
 
-        # If service is tcpwrapped we should consider the port closed
-        if service.service == "tcpwrapped":
-            continue
+            # If service is tcpwrapped we should consider the port closed
+            if service.service == "tcpwrapped":
+                continue
 
-        ip_port = IPPort(
-            address=ip.reference,
-            protocol=Protocol(protocol),
-            port=port,
-            state=PortState(service.state),
-        )
-        yield ip_port
+            ip_port = IPPort(
+                address=ip.reference,
+                protocol=Protocol(protocol),
+                port=port,
+                state=PortState(service.state),
+            )
+            yield ip_port
 
-        service_name = service.service
-        if port == 80:
-            service_name = "http"
-        if port == 443:
-            service_name = "https"
+            service_name = service.service
+            if port == 80:
+                service_name = "http"
+            if port == 443:
+                service_name = "https"
 
-        port_service = Service(name=service_name)
-        yield port_service
+            port_service = Service(name=service_name)
+            yield port_service
 
-        ip_service = IPService(ip_port=ip_port.reference, service=port_service.reference)
-        yield ip_service
+            ip_service = IPService(ip_port=ip_port.reference, service=port_service.reference)
+            yield ip_service
 
 
 def run(normalizer_meta: NormalizerMeta, raw: Union[bytes, str]) -> Iterator[OOI]:
+    """Decouple and parse Nmap XMLs and yield relevant network."""
+    # Multiple XMLs are concatenated through "\n\n". XMLs end with "\n"; we split on "\n\n\n".
+    raw = raw.decode().split("\n\n\n")
 
-    parsed: NmapReport = NmapParser.parse_fromstring(raw.decode())
+    # Relevant network object is received from the normalizer_meta.
+    network = Network(name=normalizer_meta.raw_data.boefje_meta.arguments["input"]["network"]["name"])
+    yield network
 
-    for host in parsed.hosts:
-        yield from get_ports_and_service(host)
+    logging.info(f"Parsing {len(raw)} Nmap-xml(s) for {network}.")
```

Comment:
Please use logging format instead (`logging.info("Parsing %d ... for %s", len(raw), network)` to be consistent with the rest of the codebase. (https://github.com/globality-corp/flake8-logging-format#violations-detected)

Reply:
Can you elaborate on this? `G004` in your reference only holds for non-Python3.6+. Besides, I think the `pre-commit` setup currently requires [`flynt`](https://pypi.org/project/flynt/)

Reply:
- `G004` mentions that the *first* argument should not be pre-formatted
- `flynt` will only warn about old-style formatting, it doesn't relate to the log functions using a format string as a first argument and (optionally) positional and keyword arguments to supply the format string
- I doubted whether to post this link or not because lacks proper reasoning and motivation, but in general it's safer to do it this way and (sometimes, depending on configurations such as log levels) better performing (due to lazy loading instead of pre-formatting)

Reply:
So your proposal is to use:
```python
logging.info("Parsing {n_xmls} Nmap-xml(s) for {network}.", extra={"n_xmls": len(raw), "network": network})
```
?
I don't mind changing it, though can you motivate why this (or using the old-style formatting) would be safer?

Reply:
Well, I am not really convinced after reading through that issue. I think I will stay in the f-string camp for a  while. Nevertheless, I did update the line with your suggestion :)

Reply:
I'm a huge fan of f-strings and not discouraging you to use them. But in this specific example, where safety is *not* the main argument, it's better to use the log formatting option rather than f-strings.